src/types.ts

```diff
@@ -1,15 +1,277 @@
+    : T extends typeof GET_MANY_REFERENCE
+    ? ApiPlatformAdminGetManyReferenceParams
+    : T extends typeof UPDATE
+    ? ApiPlatformAdminUpdateParams
+    : T extends typeof UPDATE_MANY
+    ? ApiPlatformAdminUpdateManyParams
+    : T extends typeof CREATE
+    ? ApiPlatformAdminCreateParams
+    : T extends typeof DELETE
+    ? ApiPlatformAdminDeleteParams
+    : T extends typeof DELETE_MANY
+    ? ApiPlatformAdminDeleteManyParams
+    : never;
+
+export interface HydraDataProviderFactoryParams {
+  entrypoint: string;
+  httpClient?: (
+    url: URL,
+    options?: HydraHttpClientOptions,
+  ) => Promise<HydraHttpClientResponse>;
+  apiDocumentationParser?: (
+    entrypointUrl: string,
+    options?: ApiDocumentationParserOptions,
+  ) => Promise<ApiDocumentationParserResponse>;
+  mercure?: Partial<MercureOptions>;
+  useEmbedded?: boolean;
+  disableCache?: boolean;
+}
 
 export interface ApiPlatformAdminDataProvider extends DataProvider {
+  getList: <RecordType extends ApiPlatformAdminRecord>(
+    resource: string,
+    params: ApiPlatformAdminGetListParams,
+  ) => Promise<GetListResult<RecordType>>;
+  getOne: <RecordType extends ApiPlatformAdminRecord>(
+    resource: string,
+    params: ApiPlatformAdminGetOneParams,
+  ) => Promise<GetOneResult<RecordType>>;
+  getMany: <RecordType extends ApiPlatformAdminRecord>(
+    resource: string,
+    params: ApiPlatformAdminGetManyParams,
+  ) => Promise<GetManyResult<RecordType>>;
+  getManyReference: <RecordType extends ApiPlatformAdminRecord>(
+    resource: string,
+    params: ApiPlatformAdminGetManyReferenceParams,
+  ) => Promise<GetManyReferenceResult<RecordType>>;
+  update: <RecordType extends ApiPlatformAdminRecord>(
+    resource: string,
+    params: ApiPlatformAdminUpdateParams,
+  ) => Promise<UpdateResult<RecordType>>;
+  updateMany: (
+    resource: string,
+    params: ApiPlatformAdminUpdateManyParams,
+  ) => Promise<UpdateManyResult>;
+  create: <RecordType extends ApiPlatformAdminRecord>(
+    resource: string,
+    params: ApiPlatformAdminCreateParams,
+  ) => Promise<CreateResult<RecordType>>;
+  delete: <RecordType extends ApiPlatformAdminRecord>(
+    resource: string,
+    params: ApiPlatformAdminDeleteParams,
+  ) => Promise<DeleteResult<RecordType>>;
+  deleteMany: (
+    resource: string,
+    params: ApiPlatformAdminDeleteManyParams,
+  ) => Promise<DeleteManyResult>;
   introspect: (
-    resource?: any,
-    options?: any,
-    reducer?: {
-      action: 'INTROSPECT';
-    },
+    _resource?: string,
+    _params?: object,
   ) => Promise<{
-    customRoutes: CustomRoutes;
-    data: { resources: unknown };
+    data: Api;
+    customRoutes?: CustomRoutes;
   }>;
-  subscribe: (...args: any) => void;
+  subscribe: (
+    resourceIds: string[],
+    callback: (document: ApiPlatformAdminRecord) => void,
+  ) => Promise<{ data: null }>;
+  unsubscribe: (
+    _resource: string,
+    resourceIds: string[],
+  ) => Promise<{ data: null }>;
```

Comment:
why not use `Promise<void>`?

Reply:
Because React-Admin data provider methods need to return something in order to be considered successful.